### ISSUE TITLE
rebuild via token and project links are not working anymore

### ISSUE THREAD

**Issue by adrianschroeter** [openSUSE/open-build-service#11035]
I got reports that using the rebuild token broke for builds where the packages are derived via project links are not working anymore since some days.

Please verify in test cases that backend gets a rebuild trigger in such situations after last weeks refactoring

(no more regression label?)

**Comment by vpereira** [comment]
Hi @adrianschroeter 

could you explain better what are the relationship between the user and "the packages derived via project links" ? What is the exactly use case that we have to take into account?

**Comment by ggardet** [comment]
The rebuild token `_obs_arm_rebuild` is not functional anymore on `openSUSE:Factory:ARM` which is a project link to `openSUSE:Factory`.

**Comment by adrianschroeter** [comment]
On Montag, 26. April 2021, 10:25:12 CEST Victor Pereira wrote:
> Hi @adrianschroeter 
> 
> could you explain better what are the relationship between the user and "the packages derived via project links" ? What is the exactly use case that we have to take into account?

just look at any architecture porting project, like openSUSE:Factory:ARM

The sources live in openSUSE:Factory, but the build results are in openSUSE:Factory:ARM .
There is the "_obs_arm_rebuild" user in :ARM, which provides the token and permission
to do rebuilds.

however, that particular token has no limitation of a package, so it should work in
any case, no matter where the package lives (but it seems not anymore). 

-- 

Adrian Schroeter ***@***.***>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

**Comment by vpereira** [comment]
> however, that particular token has no limitation of a package, so it should work in any case, no matter where the package lives (but it seems not anymore).

Today, the particular token can rebuild package everywhere, where the user has permissions to create packages.. otherwise every user would be able to rebuild every single package on obs.. so maybe you could help us to create a better authorization rule for that particular use case?

**Comment by adrianschroeter** [comment]
On Montag, 26. April 2021, 10:51:42 CEST Victor Pereira wrote:
> > however, that particular token has no limitation of a package, so it should work in any case, no matter where the package lives (but it seems not anymore).
> 
> Today, the particular token can rebuild package everywhere, where the user has permissions to create packages.. otherwise every user would be able to rebuild every single package on obs.. so maybe you could help us to create a better authorization rule for that particular use case?

Not sure that I understand, being able to build anywhere where the user has the permission is the use case here.
Therefore this user got created for this purpose.




-- 

Adrian Schroeter ***@***.***>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

**Comment by hennevogel** [comment]
@ggardet can you please explain what exactly you are trying to to? How do you trigger the token?

**Comment by ggardet** [comment]
> @ggardet can you please explain what exactly you are trying to to? How do you trigger the token?

I run `curl -d "project=openSUSE:Factory:ARM&package=$pkg" -H "Authorization: Token $MY_TOKEN" -X POST https://build.opensuse.org/trigger/rebuild` and this should trigger the rebuild of `$pkg`

**Comment by hennevogel** [comment]
@ggardet and how is this failing? You get an error or it simply doesn't rebuild?

**Comment by ggardet** [comment]
> @ggardet and how is this failing? You get an error or it simply doesn't rebuild?

I get:
```
Trigger rebuild of XYZ
<status code="ok">
  <summary>Ok</summary>
</status>
```
But no rebuild happens.

**Comment by ggardet** [comment]
It does not work on `openSUSE:Backports:SLE-15-SP3` anymore as well, so not sure it is related to project links. It looks like a general issue.

**Comment by hennevogel** [comment]
@ggardet are you okay with me manually triggering the rebuild for the package you tried today? Like OpenImageIO or coreutils

**Comment by ggardet** [comment]
> @ggardet are you okay with me manually triggering the rebuild for the package you tried today? Like OpenImageIO or coreutils

Yes, OK.

**Comment by hennevogel** [comment]
```ruby
[4] pry(main)> Backend::Api::Sources::Package.rebuild('openSUSE:Factory:ARM', 'OpenImageIO')
[backend] POST: /build/openSUSE:Factory:ARM?cmd=rebuild&package=OpenImageIO
[5] pry(main)> @package.rebuild(package: 'OpenImageIO', project: 'openSUSE:Factory:ARM')
[backend] POST: /build/openSUSE:Factory:ARM?arch=&cmd=rebuild&package=OpenImageIO&repository=
[6] pry(main)> Backend::Api::Sources::Package.rebuild('openSUSE:Factory:ARM', 'OpenImageIO', { repository: nil } )
[backend] POST: /build/openSUSE:Factory:ARM?cmd=rebuild&package=OpenImageIO&repository=
request took 0.007107124 0.13801046299999997
Backend::Error: <status code="400">
  <summary>repoid is empty</summary>
</status>

from /srv/www/obs/api/app/lib/backend/connection.rb:127:in `handle_response'

```

**Comment by ggardet** [comment]
@vpereira You closed it, but it is still broken. I now get an error message:
```
<status code="create_token/rebuild_not_authorized">
  <summary>You are not authorized to create this Token/rebuild.</summary>
</status>
```

**Comment by vpereira** [comment]
> @vpereira You closed it, but it is still broken. I now get an error message:
> 
> ```
> <status code="create_token/rebuild_not_authorized">
>   <summary>You are not authorized to create this Token/rebuild.</summary>
> </status>
> ```

Actually it was closed automatically by github..

**Comment by hennevogel** [comment]
Fixed in #11053
